File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> m = new HashMap<>();
        int res = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (m.containsKey(c)) {
                left = Math.max(left, m.get(c)+1);
            }
            res = Math.max(res, right-left+1);
            m.put(c, right);
        }
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int low = 0;
        int high = 0;
        int res = 0;
        vector<int> chars(128);
        while (high < s.size()) {
            if (chars[s[high]] > 0) {
                chars[s[low++]]--;
            } else {
                chars[s[high]]++;
                res = max(res, high++-low+1); 
            }
        }
        return res;
    }
};